executor/foreign_key.go

```diff
@@ -655,6 +660,43 @@ func GenCascadeDeleteSQL(schema, table, idx model.CIStr, fk *model.FKInfo, fkVal
 		buf.WriteString(idx.L)
 		buf.WriteString("`)")
 	}
+	err := genCascadeSQLWhereCondition(buf, fk, fkValues)
+	if err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+// GenCascadeSetNullSQL uses to generate foreign key `SET NULL` SQL, export for test.
+func GenCascadeSetNullSQL(schema, table, idx model.CIStr, fk *model.FKInfo, fkValues [][]types.Datum) (string, error) {
```

Comment:
I'm thinking it might be good to generate statments by combining AST structures instead of writing SQL strings directly. It seems more robust to me. It might be more inconvenient to implement though.

Reply:
nice catch, I will try to use AST structures later, thanks.

Reply:
I try to generate AST (https://github.com/crazycs520/tidb/tree/fk-cascade-delete-ast) but meet some strange error, maybe using AST will have another potential problems, so I prefer keep use generate SQL here.